test/apps/iodemo/io_demo.cc

```diff
@@ -1158,8 +1192,16 @@ class DemoClient : public P2pDemoCommon {
 
         while ((total_iter < opts().iter_count) && (_status == OK)) {
             VERBOSE_LOG << " <<<< iteration " << total_iter << " >>>>";
+            // if no active servers connected to the client, the 2nd parameter
+            // of std::min() will be SIZE_MAX, so max_outstanding will be
+            // equal to "global window size - 1"
+            long max_outstanding = std::min(static_cast<size_t>
+                                            (opts().window_size - 1),
+                                            static_cast<size_t>
+                                            (opts().conn_window_size) *
```

Comment:
yes, we need to cast to unsigned value to make it `SIZE_MAX` when subtracting `1`

Reply:
but we should get here only if _active_servers.size() > 0
and also we must have opts().window_size > 0 as checked by parse_window_size

Reply:
> but we should get here only if _active_servers.size() > 0

but we have the following `if` statement in this function below:
```
if (_active_servers.empty()) {
```
so, we should handle the case when `_active_servers.size() == 0`

> and also we must have opts().window_size > 0 as checked by parse_window_size

yes, we consider this here

Reply:
btw, I'd drop the global window option. Since it doesn't make sense anymore, we could express `-w` through `-a` - `w = a * num_servers`.
and `-w` is a more intuitive name for a window size argument
wdyt?

Reply:
1. active_servers.size() indeed can be 0, but according to "parse_window_size" opts().window_size is >0
2. in case the number of connections is big we may want a smaller number of total outstanding operations. that's why we have both window per connection and also shared window

Reply:
1. Ok, and that's why we need to have static_cast for `opts().conn_window_size`, since when  `_actice_servers.size() == 0` is subtractied by `1`, the 2nd parameter will be `-1`. So, `min(0, -1)` is `-1` and it's not what we want.
2. I see, good point, thank you for the explanation!

Reply:
1. can we move the call to wait_for_responses() to after if (_active_servers.empty()) { continue } ?
then we can be sure the number of active servers is >0

Reply:
> then we can be sure the number of active servers is >0

yes, makes sense

> 1. can we move the call to wait_for_responses() to after if (_active_servers.empty()) { continue } ?

do we need a progress UCP worker to get CM events needed for connection establishment with servers?
if yes, it will be problematic, since CM events won't be received

Reply:
> > 1. can we move the call to wait_for_responses() to after if (_active_servers.empty()) { continue } ?
> 
> do we need a progress UCP worker to get CM events needed for connection establishment with servers?
> if yes, it will be problematic, since CM events won't be received

discussed with @yosefe offline, it should work, since we don't progress UCP worker even before this change when the global window size is too big
@yosefe thank you, done